### ISSUE TITLE
Files simulated within rule-tester may interact with each other

### ISSUE THREAD

**Issue by rubiesonthesky** [typescript-eslint/typescript-eslint#7248]
### Suggestion

Running eslint-plugin tests in random order, I noticed that one test can fail. Command to run in random order `jest --coverage --logHeapUsage --randomize --showSeed` and reproduce the error `jest --coverage --logHeapUsage --randomize --showSeed --seed -1465958297`

It seems that this test case relies some other test to run before it

```
 FAIL  tests/rules/no-misused-promises.test.ts (477 MB heap size)
  ● no-misused-promises › invalid › 
type O = {
  func: () => void;
};
const g = async () => 'foo';
const Component = (obj: O) => null;
<Component func={g} />;
      

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      1
    Received:
      0
    
    Message:
      Should have 1 error but had 0: []

      505 |     }
      506 |     else {
    > 507 |         node_assert_1.default.strictEqual(messages.length, item.errors.length, node_util_1.default.format('Should have %d error%s but had %d: %s', item.errors.length, item.errors.length === 1 ? '' : 's', messages.length, node_util_1.default.inspect(messages)));
          |                               ^
      508 |         const hasMessageOfThisRule = messages.some(m => m.ruleId === ruleName);
      509 |         for (let i = 0, l = item.errors.length; i < l; i++) {
      510 |             const error = item.errors[i];

      at RuleTester.strictEqual (../rule-tester/dist/RuleTester.js:507:31)
      at Object.call (../rule-tester/dist/RuleTester.js:233:115)
```

It seems also happen without randomization, if I move that test case after this test case

```ts
    {
      code: `
type O = { f: () => void };
const obj: O = {
  async f() {
    return 0;
  },
};
      `,
      errors: [
        {
          line: 4,
          messageId: 'voidReturnProperty',
        },
      ],
    },
```

Not sure if that should happen or not, I can't figure out why it would work this way and I fail to see if this is intentional. But it's hard to know, so I'm reporting it :)

**Comment by JoshuaKGoldberg** [comment]
Lovely. I repro this too and have no clue why it's happening. Nice spot - I hope it wasn't too annoying to come across!

**Comment by Josh-Cena** [comment]
Here's a simplified repro:

```ts
ruleTester.run('no-misused-promises', rule, {
  valid: [],

  invalid: [
    {
      code: `
        type O = { f: () => void };
        const o: O = { f: async () => 'foo' };
      `,
      errors: [
        {
          line: 3,
          messageId: 'voidReturnProperty',
        },
      ],
    },
    {
      code: `
        type O = {
          func: () => void;
        };
        const g = async () => 'foo';
        const Component = (obj: O) => null;
        <Component func={g} />;
      `,
      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
      errors: [
        {
          line: 7,
          messageId: 'voidReturnAttribute',
        },
      ],
    },
  ],
});
```

In this case, the second test fails because it receives no errors.

Some observations:

1. If the first test reports something other than `voidReturnProperty`, then the second completes successfully (i.e. reports a lint error; same below).
2. If another test is run between these two, then all of them complete successfully. However, tests run before or after these two don't impact the result.
3. Swapping these two causes them to both succeed.
4. Adding `parserOptions.ecmaFeatures.jsx: true` to the first test causes them to both succeed.

I've located that in this particular repro, when we call `checker.getContextualType(expressionContainer)` on the JSX attribute value (`g`), the contextual type is `undefined`. However, in normal cases, it should return a contextual type corresponding to `func: () => void`. I'm not sure what's going on.

**Comment by Josh-Cena** [comment]
I think I understand what's going on, but probably someone else can confirm my theory.

Our rule tester simulates two files: `file.ts` and `react.tsx` and reuses them for each test case, using the one depending on JSX feature presence. However, the parser and the underlying TS program is reused for the entire linter run, which only gets created once per `RuleTester`. This means the two files, unless they are modules, can interact with each other because they are in script mode! In this case, because they declared a `type O` that's incompatible, TS would actually have a type checking failure on the second case. Inserting another test in between fixes this because it modifies the file content and eliminates the conflict.

Therefore, you can verify this by adding `export {};` to both test cases and see that now they work reliably!

@bradzacher (since you wrote the tester): do you think we should force all test cases to be in module mode, or do you think we should simply empty both files after each test case?

**Comment by bradzacher** [comment]
I don't think you can force TS to consider a file as a module without adding import/export syntax?

And we don't want to add code to the tests automatically as this can change the test results.

We probably want to "null out" the file used by a test after the test finishes to prevent it influencing other tests. However this is probably going to slow things down because TS will do another typecheck cycle.

Perhaps we should instead keep track of all the files and ensure we null them out when a different filename is used?

Or perhaps we could use one program instance per filename and keep a map of them to reuse them properly?

**Comment by Josh-Cena** [comment]
> I don't think you can force TS to consider a file as a module without adding import/export syntax?

This is what `moduleDetection: "force"` is for. Unless we have tests relying on the file format being non-module or the ability to toggle module format I think this is the way to go.
